scenario-compiler/src/main/scala/pl/touk/nussknacker/engine/extension/Cast.scala

```diff
@@ -0,0 +1,117 @@
+package pl.touk.nussknacker.engine.extension
+
+import cats.data.ValidatedNel
+import cats.implicits.catsSyntaxValidatedId
+import pl.touk.nussknacker.engine.api.generics.{GenericFunctionTypingError, MethodTypeInfo, Parameter}
+import pl.touk.nussknacker.engine.api.typed.typing
+import pl.touk.nussknacker.engine.api.typed.typing.{Typed, TypedObjectWithValue, TypingResult, Unknown}
+import pl.touk.nussknacker.engine.definition.clazz.{ClassDefinitionSet, FunctionalMethodDefinition, MethodDefinition}
+import pl.touk.nussknacker.engine.extension.CastMethodDefinitions._
+import pl.touk.nussknacker.engine.util.Implicits.RichScalaMap
+
+import scala.util.Try
+
+sealed trait Cast {
+  def canCastTo(className: String): Boolean
+  def castTo[T](className: String): T
+  def castToOrNull[T >: Null](className: String): T
+}
+
+class CastImpl(target: Any, classLoader: ClassLoader) extends Cast {
+
+  override def canCastTo(className: String): Boolean =
+    classLoader.loadClass(className).isAssignableFrom(target.getClass)
+
+  override def castTo[T](className: String): T = {
+    val clazz = classLoader.loadClass(className)
+    if (clazz.isInstance(target)) {
+      clazz.cast(target).asInstanceOf[T]
+    } else {
+      throw new ClassCastException(s"Cannot cast: ${target.getClass} to: $className")
+    }
+  }
+
+  override def castToOrNull[T >: Null](className: String): T = Try { castTo[T](className) }.getOrElse(null)
+}
+
+private[extension] object CastImpl extends ExtensionMethodsImplFactory {
+  override def create(target: Any, classLoader: ClassLoader): Any =
+    new CastImpl(target, classLoader)
+}
+
+private[extension] class CastMethodDefinitions(private val classesWithTyping: Map[Class[_], TypingResult]) {
+
+  def createDefinitions(clazz: Class[_]): Map[String, List[MethodDefinition]] =
+    classesWithTyping.filterKeysNow(c => isAssignable(clazz, c)) match {
+      case allowedClasses if allowedClasses.isEmpty => Map.empty
+      case allowedClasses                           => definitions(allowedClasses)
+    }
+
+  private def isAssignable(clazz: Class[_], targetClazz: Class[_]): Boolean =
```

Comment:
I think that inlining this method would give more benefits. I was confused, because this method doesn't check that class is assignable but also that it is not the same class. Instead of this method, I would create a local variable `childTypes` to explain why we do this filtering.

Reply:
Done